Replace ObserveSurfaceIntegrals with ObserveTimeSeriesOnSurface

markscheel opened the pull request:
## Proposed changes

ObserveSurfaceIntegrals was able to observe only integrals.

ObserveSurfaceIntegrals has been upgraded to be able to observe the results of arbitrary ComputeItems on the surface (if the type is observable by WriteReductionData),
and has been renamed to ObserveTimeSeriesOnSurface.

To observe integrals, one specifies a new ComputeItem that does the integral.
But now one can observe things like Christodoulou mass.

There are three commits:
1) Separate ApparentHorizons/Tags.hpp into Tags.hpp and TagsDeclarations.hpp
2) Add ComputeItems necessary for doing integrals
3) Replace ObserveSurfaceIntegrals with ObserveTimeSeriesOnSurface.

This addresses one of my comments on #1249

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Replace ObserveSurfaceIntegrals with ObserveTimeSeriesOnSurface.

Now the things being observed can be the result of arbitrary
ComputeItems (that return a type that is observable by WriteReductionData).
To do integrals, use the new ComputeItems that do the integrals.
So now it is possible, for instance, to observe things
like the Christodoulou Mass on an apparent horizon.

Also simplified some of the logic, eliminated hardcoding of doubles,
and eliminated the need for complicated indexing.

markscheel committed:
Split ApparentHorizons/Tags.hpp

It is now split into 3 files:
 - TagsDeclarations.hpp (declares all the tags)
 - Tags.hpp (defines all the tags)
 - TagsTypeAliases.hpp (defines type aliases used by Tags.hpp)

In TagsTypeAliases.hpp, the type aliases are now in a namespace
StrahlkorperTags::aliases instead of the previous StrahlkorperTags_detail.
This necessitates renaming changes in StrahlkorperGr.?pp and
in Tags.cpp.

markscheel committed:
Add SurfaceIntegral and AreaElement ComputeItems.

geoffrey4444 reviewed (changes requested):
Travis is sad

markscheel commented:
Attempted to make travis happy. Squashed.

kidder commented on src/ApparentHorizons/TagsDeclarations.hpp:
```diff
@@ -0,0 +1,70 @@
+namespace StrahlkorperTags {
+
+namespace StrahlkorperTags_detail {
+// Shorter names for longer types used below.
```
I don't see half of these names below

markscheel commented:
Hopefully IWYU is happy now.  It sometimes gives different output on Travis than on my local container.

wthrowe commented on src/ApparentHorizons/TagsDeclarations.hpp:
```diff
@@ -0,0 +1,76 @@
+
+namespace StrahlkorperTags {
+
+namespace StrahlkorperTags_detail {
```
Are these details needed in the declarations file, or can they be moved to Tags.hpp?

markscheel replied:
Here's why I did this. Before I made this change, I had the following problem:

 - StrahlkorperGr.hpp depends on Tags.hpp  because it uses type aliases in StrahlkorperTags_detail (defined in Tags.hpp)

 - I want to add ComputeItems to Tags.hpp, which will make Tags.hpp depend on StrahlkorperGr.hpp (which declares the functions used by those ComputeItems).

So how to break the circular dependency?  A few options:

1.  Put `StrahlkorperTags_detail` into its own hpp file, so it can be included into Tags.hpp and StrahlkorperGr.hpp.

1.  Same as previous but also put the Tags.hpp declarations into the same hpp file with `Strahlkorper_tags_detail`, since at some point we will want only the declarations and not the definitions.  This is what I did (but I see now why it is confusing).

1. Have three different files: Tags.hpp, TagsDeclarations.hpp, and TagsTypeAliases.hpp, where the last file includes what is now in `StrahlkorperTags_detail` (and put into a namespace without 'detail' in the name).

1. Remove the usage of `Strahlkorper_tags_detail` in StrahlkorperGr.hpp by explicitly specifying the long types instead of using type aliases.  Then `Strahlkorper_tags_detail` is used only in Tags.hpp.  But we need to write out four-line type names in StrahlkorperGr.hpp.

Preferences? Option 3 seems cleanest to me. (or some other suggestion)?

hrueter commented on src/ApparentHorizons/Tags.hpp:
```diff
@@ -238,3 +212,32 @@ using compute_items_tags =
+/// \ingroup SurfacesGroup
+/// Holds tags and ComputeItems associated with a `::Strahlkorper` that
+/// also need a metric.
+namespace StrahlkorperGrTags {
```
To be consistent with how we name other `Tags`-namespaces:
`StrahlkorperGr::Tags`

hrueter commented on src/NumericalAlgorithms/Interpolation/Callbacks/ObserveTimeSeriesOnSurface.hpp:
```diff
@@ -51,113 +37,76 @@ namespace intrp {
 
 template <typename... Ts>
-struct reduction_data_type<tmpl::list<Ts...>> {
+struct reduction_data_tag_type<tmpl::list<Ts...>> {
```
FYI: PR #1269 deletes this struct

hrueter reviewed (commented):
some comments from @nilsleiffischer and @hrueter

markscheel commented:
The first commit ("Split ApparentHorizons/Tags.hpp") has been changed in accordance with above requests/comments, and now there are 3 files Tags.hpp, TagsDeclarations.hpp, and TagsTypeAliases.hpp.  Because this change involved renaming a few things and thus touched many files, I felt it would be easier to review if I squashed fixups to the first commit, so I did so.  Also I rebased onto develop when squashing.

kidder commented:
@wthrowe @geoffrey4444 please look at

markscheel commented:
The failing travis test is another one of those gccdebug tests where a large number of tests segfault for no apparent reason.

markscheel commented:
ping @sxs-collaboration/spectre-core-devs
